### PR TITLE
systemd: unit hardening

### PR DESCRIPTION
**Pull request by mweinelt**
This was part of my NixOS packaging effort, which I'm hereby upstreaming.

- Switches away from having a dedicated, persistent user in favor of
DynamicUsers.
- Lock down what the service can ultimately do. On systemd 246 that gets
us:
  → Overall exposure level for url-bot-rs.service: 1.3 OK 🙂

I only tested this with a very simple configuration, so it may well be that I missed things. Especially situation where a database that is not in-memory might be used.

### CODE REVIEW

**Comment by mweinelt** [comment]
We could add `StateDirectory=url-bot-rs` for good measure. That would be the place to store the sqlite.db at.

**Comment by nuxeh** [comment]
Thanks! I'll give this a test with the deb package install. But something that's been waiting around for a while!

Fixes #89

File: systemd/url-bot-rs.service [hunk 1]
```diff
@@ -4,10 +4,38 @@ After=network.target
...
 Restart=always
 RestartSec=10
 
+AmbientCapabilities=
+CapabilityBoundingSet=
+DeviceAllow=/dev/stderr
+DeviceAllow=/dev/stdin
+DeviceAllow=/dev/stdout
+DevicePolicy=strict
+DynamicUser=true
```
Comment (polyzen) [inline systemd/url-bot-rs.service:18]: This is set twice.

**Comment by polyzen** [comment]
Lol I really need to switch the box I'm using for this from CentOS 7. Quite a few of these hardening options don't exist in systemd 219 😅.

Evidently the `getrandom` system call is also not available there, so had to add:
```
DeviceAllow=/dev/random
DeviceAllow=/dev/urandom
```

**Comment by mweinelt** [comment]
Thanks, fixed.

**Comment by nuxeh** [comment]
I got around to testing on Debian too, and the service doesn't start. I think due to the location of the state directory for dynamic user, but I haven't got to the bottom of it yet.

Added a github workflow to test the service start though, recently.

**Comment by mweinelt** [comment]
> I think due to the location of the state directory for dynamic user, but I haven't got to the bottom of it yet.

The state directory on every start of the unit will be mounted to `/var/lib/url-bot-rs`

**Comment by nuxeh** [comment]
Looks good... Thanks!

**Comment by nuxeh** [comment]
I was planning to make a new release, but it seems there are some issues with `DynamicUser` for Debian package based upgrades, which I've outlined in #369 .

And also, the service start workflow didn't wait long enough to check. Wait some additional time and it fails, on the Ubuntu runner at least. I'm not sure why  but it's getting sent SIGSYS here.

```
-- Logs begin at Wed 2020-12-16 15:03:59 UTC, end at Wed 2020-12-16 17:24:26 UTC. --
Dec 16 17:24:15 fv-az32-410 systemd[1]: /lib/systemd/system/url-bot-rs.service:25: Unknown lvalue 'ProtectClock' in section 'Service'
Dec 16 17:24:15 fv-az32-410 systemd[1]: /lib/systemd/system/url-bot-rs.service:28: Unknown lvalue 'ProtectHostname' in section 'Service'
Dec 16 17:24:15 fv-az32-410 systemd[1]: /lib/systemd/system/url-bot-rs.service:29: Unknown lvalue 'ProtectKernelLogs' in section 'Service'
Dec 16 17:24:15 fv-az32-410 systemd[1]: /lib/systemd/system/url-bot-rs.service:38: Unknown system call group, ignoring: @system-service
Dec 16 17:24:15 fv-az32-410 systemd[1]: Started url-bot-rs.
Dec 16 17:24:16 fv-az32-410 systemd[1]: url-bot-rs.service: Main process exited, code=killed, status=31/SYS
Dec 16 17:24:16 fv-az32-410 systemd[1]: url-bot-rs.service: Failed with result 'signal'.
Dec 16 17:24:26 fv-az32-410 systemd[1]: url-bot-rs.service: Service hold-off time over, scheduling restart.
Dec 16 17:24:26 fv-az32-410 systemd[1]: url-bot-rs.service: Scheduled restart job, restart counter is at 1.
Dec 16 17:24:26 fv-az32-410 systemd[1]: Stopped url-bot-rs.
Dec 16 17:24:26 fv-az32-410 systemd[1]: Started url-bot-rs.
Dec 16 17:24:26 fv-az32-410 systemd[1]: url-bot-rs.service: Main process exited, code=killed, status=31/SYS
Dec 16 17:24:26 fv-az32-410 systemd[1]: url-bot-rs.service: Failed with result 'signal'.
```

**Comment by mweinelt** [comment]
Maybe try disabling or relaxing the SystemCallFilter.

**Pull request merged**